contracts/Staking.sol

```diff
@@ -0,0 +1,507 @@
+     * @param _amount The amount of tokens to be locked
+     * @param _manager The manager entity for this particular lock. This entity will have full control over the lock, in particular will be able to unlock it
+     * @param _data Data to parametrize logic for the lock to be enforced by the manager
+     * @return The id of the newly created lock
+     */
+    function lock(uint256 _amount, address _manager, bytes _data) external returns (uint256) {
+        // locking 0 tokens is invalid
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // check enough unlocked tokens are available
+        require(_amount <= unlockedBalanceOf(msg.sender), ERROR_NOT_ENOUGH_BALANCE);
+
+        accounts[msg.sender].lastLockId++;
+        uint256 _lockId = accounts[msg.sender].lastLockId;
+        accounts[msg.sender].activeLockIds.push(_lockId);
+        Lock storage lock_ = accounts[msg.sender].locks[_lockId];
+        lock_.amount = _amount;
+        lock_.unlockedAt = MAX_UINT64;
+        lock_.manager = ILockManager(_manager);
+        lock_.data = _data;
+
+        emit Locked(msg.sender, _lockId, _amount, _manager, _data);
+
+        return _lockId;
+    }
+
+    /**
+     * @notice Try to unlock as much locks belonging to `_account` as possible
+     * @dev It won't work (it will revert) if one of the managers is an EOA
+     * @param _account Owner whose locks are to be unlocked
+     */
+    function unlockAll(address _account) external {
+        Account storage account = accounts[_account];
+
+        for (uint256 i = account.activeLockIds.length; i > 0; i--) {
+            if (canUnlock(_account, account.activeLockIds[i - 1])) {
+                unlock(_account, account.activeLockIds[i - 1]);
+            }
+        }
+    }
+
+    /**
+     * @notice Try to unlock all locks belonging to `_account` and revert if any of them fail
+     * @param _account Owner whose locks are to be unlocked
+     */
+    function unlockAllOrNone(address _account) external {
+        Account storage account = accounts[_account];
+
+        for (uint256 i = account.activeLockIds.length; i > 0; i--) {
+            unlock(_account, account.activeLockIds[i - 1]);
+        }
+    }
+
+    /**
+     * @notice Transfer `_amount` tokens to `_to``_toLockId > 0 ? '\'s lock #' + _toLockId : ''`
+     * @param _amount Number of tokens to be transferred
+     * @param _to Recipient of the tokens
+     * @param _toLockId Lock id of the recipient to add the tokens to, if any
+     */
+    function transfer(uint256 _amount, address _to, uint256 _toLockId) external {
+        // transfererring 0 staked tokens is invalid
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+        // have enough unlocked funds
+        require(_amount <= unlockedBalanceOf(msg.sender), ERROR_NOT_ENOUGH_BALANCE);
+
+        // first valid lock starts at 1, so _toLockId = 0 means no lock
+        if (_toLockId > 0) {
+            _updateActiveLockAmount(_to, _toLockId, _amount, true);
```

Comment:
If I'm not wrong this function allows any account to transfer some unlocked amount in their power to a recipient account and add that amount to an existing lock of the recipient account if they want, i.e. any account could lock the transferring amount of a recipient on a transfer, is that ok?

Reply:
Yes, that's correct. The idea is that the account transfers the tokens under the condition that they are locked. I think the use case is to simplify UX allowing to do certain operations in one transaction instead of two.
Do you see any potential harm in it?

Reply:
Not at all, I was asking to check if I understood it correctly :)